src/consensus/replay_tower.zig

```diff
@@ -1579,6 +1613,622 @@ fn optimisticallyBypassVoteStakeThresholdCheck(
+                try vote_slots.put(vote.slot);
+            }
+        }
+        if (vote_state.root_slot) |root| {
+            const vote = Lockout{ .slot = root, .confirmation_count = MAX_LOCKOUT_HISTORY };
+            try vote_slots.put(vote.slot);
+        }
+
+        // The last vote in the vote stack is a simulated vote on bank_slot, which
+        // we added to the vote stack earlier in this function by calling processVote().
+        // We don't want to update the ancestors stakes of this vote b/c it does not
+        // represent an actual vote by the validator.
+
+        // Note: It should not be possible for any vote state in this bank to have
+        // a vote for a slot >= bank_slot, so we are guaranteed that the last vote in
+        // this vote stack is the simulated vote, so this fetch should be sufficient
+        // to find the last unsimulated vote.
+        std.debug.assert(
+            if (vote_state.nthRecentLockout(0)) |l| l.slot == bank_slot else false,
+        );
+
+        if (vote_state.nthRecentLockout(1)) |vote| {
+            // Update all the parents of this last vote with the stake of this vote account
+            try updateAncestorVotedStakes(
+                allocator,
+                &voted_stakes,
+                vote.slot,
+                voted_stake,
+                ancestors,
+            );
+        }
+        total_stake += voted_stake;
+    }
+
+    try populateAncestorVotedStakes(
+        allocator,
+        &voted_stakes,
+        vote_slots.items(),
+        ancestors,
+    );
+
+    // As commented above, since the votes at current bank_slot are
+    // simulated votes, the voted_stake for `bank_slot` is not populated.
+    // Therefore, we use the voted_stake for the parent of bank_slot as the
+    // `fork_stake` instead.
+    const fork_stake = blk: {
+        var bank_ancestors = ancestors.get(bank_slot) orelse break :blk @as(u64, 0);
+        var max_parent: ?Slot = null;
+        for (bank_ancestors.items()) |slot| {
+            if (max_parent == null or slot > max_parent.?) {
+                max_parent = slot;
+            }
+        }
+        if (max_parent) |parent| {
+            break :blk voted_stakes.get(parent) orelse @as(u64, 0);
+        } else {
+            break :blk @as(u64, 0);
+        }
+    };
+
+    return ComputedBankState{
+        .voted_stakes = voted_stakes,
+        .total_stake = total_stake,
+        .fork_stake = fork_stake,
+        .lockout_intervals = lockout_intervals,
+        .my_latest_landed_vote = my_latest_landed_vote,
+    };
+}
+
+pub fn lastVotedSlotInBank(
+    allocator: std.mem.Allocator,
+    accounts_db: *AccountsDB,
+    vote_account_pubkey: *const Pubkey,
+) ?Slot {
+    const vote_account = accounts_db.getAccount(vote_account_pubkey) catch return null;
+    const vote_state = stateFromAccount(
+        allocator,
+        &vote_account,
+        vote_account_pubkey,
+    ) catch return null;
```

Comment:
I don't think it is wise to allow ignoring potential allocation errors like this.
I think instead `stateFromAccount` should be made to return only `std.mem.Allocator.Error || error{BincodeError}`, avoiding the fact that bincode currently propagates a bunch of `anyerror` silliness - that would in turn allow this call site to safely do:
```zig
const vote_state = stateFromAccount(
    allocator,
    &vote_account,
    vote_account_pubkey,
) catch |err| switch (err) {
    error.OutOfMemory => |e| return e,
    error.BincodeError => return null,
};
```

Reply:
[460e2d9](https://github.com/Syndica/sig/pull/798/commits/460e2d930f60ae9794d1690c54d2dd00ec21d225)